src/graph/sampling/neighbor/neighbor_cpu.h

```diff
@@ -0,0 +1,167 @@
+    // moves valid elements to the head of arrays and create a view of the original
+    // array. The implementation consumes a little extra memory than the actual requirement.
+    //
+    // Otherwise, directly use the row and col arrays to construct the result graph.
+
+    bool all_has_fanout = true;
+    if (replace) {
+      all_has_fanout = true;
+    } else {
+#pragma omp parallel for reduction(&&:all_has_fanout)
+      for (int64_t i = 0; i < num_nodes; ++i) {
+        const IdxType nid = nodes_data[i];
+        const IdxType len = indptr[nid + 1] - indptr[nid];
+        all_has_fanout = all_has_fanout && (len >= fanout);
+      }
+    }
+
+    IdArray row = aten::Full(-1, num_nodes * fanout, sizeof(IdxType) * 8, hg->Context());
+    IdArray col = aten::Full(-1, num_nodes * fanout, sizeof(IdxType) * 8, hg->Context());
+
+    IdxType* row_data = static_cast<IdxType*>(row->data);
+    IdxType* col_data = static_cast<IdxType*>(col->data);
+
+#pragma omp parallel for
+    for (int64_t i = 0; i < num_nodes; ++i) {
+      const IdxType nid = nodes_data[i];
+      const IdxType off = indptr[nid];
+      const IdxType len = indptr[nid + 1] - off;
+      if (len <= fanout && !replace) {
+        // neighborhood size <= fanout and w/o replacement, take all neighbors
+        for (int64_t j = 0; j < len; ++j) {
+          row_data[i * fanout + j] = nid;
+          col_data[i * fanout + j] = indices[off + j];
+        }
+      } else {
+        FloatArray weight_selected = weight[etype];
+        if (weight[etype]->shape[0] != 0) {
+          weight_selected = LightSlice<IdxType>(weight[etype], adj[2], off, len);
+        }
+        IdArray chosen = choose_fn(fanout, len, weight_selected);
+        const IdxType* chosen_data = static_cast<IdxType*>(chosen->data);
+        for (int64_t j = 0; j < fanout; ++j) {
+          row_data[i * fanout + j] = nid;
+          col_data[i * fanout + j] = indices[off + chosen_data[j]];
+        }
+      }
+    }
+
+    if (!all_has_fanout) {
+      // correct the array by remove_if
+      IdxType* new_row_end = std::remove_if(row_data, row_data + num_nodes * fanout,
+                                            [] (IdxType i) { return i == -1; });
+      IdxType* new_col_end = std::remove_if(col_data, col_data + num_nodes * fanout,
+                                            [] (IdxType i) { return i == -1; });
+      const int64_t new_len = (new_row_end - row_data);
+      CHECK_EQ(new_col_end - col_data, new_len);
+      CHECK_LT(new_len, num_nodes * fanout);
+      row = row.CreateView({new_len}, row->dtype);
+      col = col.CreateView({new_len}, col->dtype);
+    }
+
+    if (dir == EdgeDir::kIn)
+      std::swap(row, col);
+
+    subrels[etype] = UnitGraph::CreateFromCOO(
+      hg->GetRelationGraph(etype)->NumVertexTypes(),
+      hg->NumVertices(src_vtype),
+      hg->NumVertices(dst_vtype),
+      row, col);
+  }
+
+  return CreateHeteroGraph(hg->meta_graph(), subrels);
+}
```

Comment:
I'm thinking of representing  neighbor sampling as a sparse matrix operation, like "for each row, randomly take K non-zero entries".

Also, we probably need a COO-based implementation as well.

Reply:
That's a good idea. Will incorporate your changes.